core/src/kura.rs

```diff
@@ -180,86 +190,156 @@ impl Kura {
-                        }
-                    }
+            // We don't want to hold up other threads so we drop the lock on the block data.
+            drop(block_data_guard);
+
+            if let Some(path) = kura.block_plain_text_path.as_ref() {
+                let mut plain_text_file = std::fs::OpenOptions::new()
+                    .create(true)
+                    .append(true)
+                    .open(path)
+                    .expect("Couldn't create file for plain text blocks.");
+
+                for new_block in &blocks_to_be_written {
+                    serde_json::to_writer_pretty(&mut plain_text_file, new_block.as_ref())
+                        .expect("Failed to write to plain text file for blocks.");
+                }
+            }
+
+            let mut block_store_guard = kura.block_store.lock();
+            if let Err(error) = block_store_guard.write_index_count(start_height as u64) {
+                error!("Failed to write index count, ERROR = {}", error);
+                panic!("Kura has encountered a fatal I/O error.");
+            }
+
+            for block in blocks_to_be_written {
+                let serialized_block: Vec<u8> = block.encode_versioned();
+
+                if let Err(error) = block_store_guard.append_block_to_chain(&serialized_block) {
+                    error!("Failed to store block, ERROR = {}", error);
+                    panic!("Kura has encountered a fatal I/O error.");
                 }
             }
         }
-        info!("Kura block thread is shutting down");
     }
 
     /// Get the hash of the block at the provided height.
     #[allow(clippy::unwrap_in_result, clippy::expect_used)]
     pub fn get_block_hash(&self, block_height: u64) -> Option<HashOf<VersionedCommittedBlock>> {
-        let hash_array_guard = self.block_hash_array.lock();
-        if block_height == 0 || block_height > hash_array_guard.len() as u64 {
+        let hash_data_guard = self.block_data.lock();
+        if block_height == 0 || block_height > hash_data_guard.len() as u64 {
             return None;
         }
         let index: usize = (block_height - 1)
             .try_into()
             .expect("block_height fits in 32 bits or we are running on a 64 bit machine");
-        Some(hash_array_guard[index])
+        Some(hash_data_guard[index].0)
     }
 
-    /// Put a block in the queue to be stored by Kura. If the queue is
-    /// full, wait until there is space.
-    pub async fn store_block_async(&self, block: VersionedCommittedBlock) {
-        if let Err(SendError(block)) = self.block_sender.send(block).await {
-            error!(?block, "unable to write block, kura thread was closed");
+    /// Search through blocks for the height of the block with the given hash.
+    pub fn get_block_height_by_hash(&self, hash: &HashOf<VersionedCommittedBlock>) -> Option<u64> {
+        let hash_data_guard = self.block_data.lock();
+
+        (*hash != Hash::zeroed().typed())
```

Comment:
It's a short circuit for the null hash. Simply put, no block can have the hash zero but we do use all zeroes to represent a fresh blockchain. So instead of searching through all the blocks then returning none we return none right away.

Reply:
I mean is there any execution path that would feed in empty hash except by mistake

Reply:
I think so. A lot of the sumeragi and block sync code is certainly written assuming being able to do so.

Reply:
The code that this replaces did this check.

Reply:
> I think so. A lot of the sumeragi and block sync code is certainly written assuming being able to do so.

being able to send empty hash doesn't make it ok to do so. Sending empty hash is equivalent to sending any other dangling hash value. But I guess @Erigara will take care of this in his PR